src/account/BiconomySmartAccountV2.ts

```diff
@@ -1513,21 +1514,62 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
       userOp.maxFeePerGas = gasFeeValues?.maxFeePerGas as Hex
       userOp.maxPriorityFeePerGas = gasFeeValues?.maxPriorityFeePerGas as Hex
 
-      userOp = await this.getPaymasterUserOp(
-        userOp,
-        buildUseropDto.paymasterServiceData
-      )
-      return userOp
+      if(buildUseropDto.gasOffset){
+        userOp = await this.estimateUserOpGas(userOp)
+
+        const {verificationGasLimitIncrement, preVerificationGasIncrement, callGasLimitIncrement, maxFeePerGasIncrement, maxPriorityFeePerGasIncrement} = buildUseropDto.gasOffset;
+        userOp.verificationGasLimit = toHex(parseInt((Number(userOp.verificationGasLimit ?? 0) * convertToFactor(verificationGasLimitIncrement)).toString()));
+        userOp.preVerificationGas = toHex(parseInt((Number(userOp.preVerificationGas ?? 0) * convertToFactor(preVerificationGasIncrement)).toString()));
+        userOp.callGasLimit = toHex(parseInt((Number(userOp.callGasLimit ?? 0) * convertToFactor(callGasLimitIncrement)).toString()));
+        userOp.maxFeePerGas = toHex(parseInt((Number(userOp.maxFeePerGas ?? 0) * convertToFactor(maxFeePerGasIncrement)).toString()));
+        userOp.maxPriorityFeePerGas = toHex(parseInt((Number(userOp.maxPriorityFeePerGas ?? 0) * convertToFactor(maxPriorityFeePerGasIncrement)).toString()));
+        
+        userOp = await this.getPaymasterUserOp(
+          userOp,
+          {...buildUseropDto.paymasterServiceData, calculateGasLimits: false},
+        )
+        return userOp
+      } else {
```

Comment:
why this else path here?
can we somehow optimise this code block

Reply:
Fixed